modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java

```diff
@@ -0,0 +1,327 @@
+
+    private static final String PARTITION_BUILD_INDEX_KEY_PREFIX = "partitionBuildIndex";
+
+    private final CatalogManager catalogManager;
+
+    private final MetaStorageManager metaStorageManager;
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean stopGuard = new AtomicBoolean();
+
+    /** Constructor. */
+    public IndexAvailabilityController(CatalogManager catalogManager, MetaStorageManager metaStorageManager, IndexBuilder indexBuilder) {
+        this.catalogManager = catalogManager;
+        this.metaStorageManager = metaStorageManager;
+
+        addListeners(catalogManager, metaStorageManager, indexBuilder);
+    }
+
+    @Override
+    public void close() {
+        if (!stopGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+    }
+
+    private void addListeners(CatalogService catalogService, MetaStorageManager metaStorageManager, IndexBuilder indexBuilder) {
+        catalogService.listen(CatalogEvent.INDEX_CREATE, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onIndexCreate((CreateIndexEventParameters) parameters).thenApply(unused -> false);
+        });
+
+        catalogService.listen(CatalogEvent.INDEX_DROP, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onIndexDrop((DropIndexEventParameters) parameters).thenApply(unused -> false);
+        });
+
+        catalogService.listen(CatalogEvent.INDEX_AVAILABLE, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onIndexAvailable((MakeIndexAvailableEventParameters) parameters).thenApply(unused -> false);
+        });
+
+        metaStorageManager.registerPrefixWatch(ByteArray.fromString(PARTITION_BUILD_INDEX_KEY_PREFIX), new WatchListener() {
+            @Override
+            public CompletableFuture<Void> onUpdate(WatchEvent event) {
+                return onUpdatePartitionBuildIndexKey(event).thenApply(unused -> null);
+            }
+
+            @Override
+            public void onError(Throwable e) {
+                LOG.error("Error on handle partition build index key", e);
+            }
+        });
+
+        indexBuilder.listen((indexId, tableId, partitionId) -> onIndexBuildForPartition(indexId, partitionId));
+    }
+
+    private CompletableFuture<?> onIndexCreate(CreateIndexEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            int indexId = parameters.indexDescriptor().id();
+
+            int partitions = getPartitionCountFromCatalog(indexId, parameters.catalogVersion());
+
+            ByteArray startBuildIndexKey = startBuildIndexKey(indexId);
```

Comment:
Looks like this key is not about 'index build start', but about 'index is being built' because the index is built strictly between the key is created and removed. So I suggest to rename the key accordingly.

Reply:
How exactly to rename?

Reply:
Like 'indexBuild.inProgress.'

Reply:
Why not